### PR TITLE
[Feature] Support FastFCN

### PR DESCRIPTION
**Pull request by MengzhangLI**
## Introduction

Support an interesting algorithm which could use smaller feature map getting good results with less computation (FLOPs), [FastFCN: Rethinking Dilated Convolution in the Backbone for SemanticSegmentation](http://wuhuikai.me/FastFCNProject/fast_fcn.pdf).

Original Repo: https://github.com/wuhuikai/FastFCN

Its framework is shown as below:
image

And core novelty of it is Joint Pyramid Upsampling (JPU):
image

## Modification

1. Add a new neck modified from https://github.com/wuhuikai/FastFCN
2. Add config files.

**Current results on Cityscapes using MMSegmentation are below:**

image

## Future work
TO DO:

- [x] Aligning training & testing metric of JPU for PSPNet, DeepLabV3 and EncNet.
- [x] Adding config files.
- [x] Making fair comparison about FPS.
- [x] Adding README and uploading models & logs.

(Optional) Results on **ADE20K** and **Pascal Context**.

### CODE REVIEW

File: mmseg/models/necks/jpu.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+        self.conv5 = nn.Sequential(
+            ConvModule(
+                self.in_channels[-1],
+                self.out_channels,
+                kernel_size=3,
+                padding=1,
+                bias=False,
+                conv_cfg=conv_cfg,
+                norm_cfg=norm_cfg,
+                act_cfg=act_cfg))
```
Comment (Junjun2016) [inline mmseg/models/necks/jpu.py:83]: Use convs.
Refer to https://github.com/open-mmlab/mmsegmentation/blob/master/mmseg/models/necks/fpn.py#L120.

File: mmseg/models/necks/jpu.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+                in_channels=3 * self.out_channels,
+                out_channels=self.out_channels,
+                kernel_size=3,
+                stride=1,
+                padding=8,
+                dilation=8,
+                dw_norm_cfg=norm_cfg,
+                dw_act_cfg=None,
+                pw_norm_cfg=norm_cfg,
+                pw_act_cfg=act_cfg))
```
Comment (Junjun2016) [inline mmseg/models/necks/jpu.py:131]: Can use `for` loop.

Comment (Junjun2016) [reply mmseg/models/necks/jpu.py:131]: Making dilation as an argument.
Refer to https://github.com/open-mmlab/mmsegmentation/blob/master/mmseg/models/decode_heads/aspp_head.py#L23.

File: mmseg/models/necks/jpu.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+                 conv_cfg=None,
+                 norm_cfg=dict(type='BN'),
+                 act_cfg=dict(type='ReLU'),
+                 init_cfg=None):
+        super(JPU, self).__init__(init_cfg=init_cfg)
+        self.in_channels = in_channels
+        self.out_channels = out_channels
+        self.align_corners = align_corners
+        self.out_indices = out_indices
+
```
Comment (Junjun2016) [inline mmseg/models/necks/jpu.py:52]: Should add some asserts for argumens.

File: mmseg/models/necks/jpu.py [hunk 4]
```diff
@@ -90,35 +95,35 @@ def __init__(self,
...
-        feats = [self.conv5(x_32), self.conv4(x_16), self.conv3(x_8)]
+        assert len(inputs) == len(self.in_channels), 'Length of inputs must \
+                                           be the same with self.in_channels!'
+
+        feats = [
+            self.conv_layers[i](inputs[i + 1])
+            for i in range(len(self.in_channels) - 1)
+        ]
+
+        _, _, h, w = feats[0].size()
```
Comment (Junjun2016) [inline mmseg/models/necks/jpu.py:110]:
```suggestion
        h, w = feats[0].shape[2:]
```

File: mmseg/models/necks/jpu.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+    Args:
+        in_channels (Tuple[int], optional): The number of input channels
+            for each convolution operations before upsampling.
+            Default: (256, 512, 1024, 2048).
+        start_level (int): Index of the start input backbone level used to
+            build the feature pyramid. Default: 1.
+        end_level (int): Index of the end input backbone level (exclusive) to
+            build the feature pyramid. Default: -1, which means the last level.
+        dilations (tuple[int]): Dilation rate of each Depthwise
+            Separable ConvModule.
```
Comment (Junjun2016) [inline mmseg/models/necks/jpu.py:28]: Missing default.

File: mmseg/models/necks/jpu.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+                 align_corners=False,
+                 conv_cfg=None,
+                 norm_cfg=dict(type='BN'),
+                 act_cfg=dict(type='ReLU'),
+                 init_cfg=None):
+        super(JPU, self).__init__(init_cfg=init_cfg)
+        assert isinstance(in_channels, tuple)
+        assert isinstance(dilations, tuple)
+        self.in_channels = in_channels
+        self.out_channels = in_channels[start_level]
```
Comment (Junjun2016) [inline mmseg/models/necks/jpu.py:58]: `out_channels` should be an argument.

File: mmseg/models/necks/jpu.py [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+        self.conv_layers = nn.ModuleList()
+        self.dilation_layers = nn.ModuleList()
+        for i in range(self.start_level, self.backbone_end_level):
+            conv_layer = nn.Sequential(
+                ConvModule(
+                    self.in_channels[i],
+                    self.out_channels,
+                    kernel_size=3,
+                    padding=1,
+                    bias=False,
```
Comment (Junjun2016) [inline mmseg/models/necks/jpu.py:80]: Can remove `bias=False`?

File: mmseg/models/necks/jpu.py [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+                    conv_cfg=conv_cfg,
+                    norm_cfg=norm_cfg,
+                    act_cfg=act_cfg))
+            self.conv_layers.append(conv_layer)
+        for i in range(len(dilations)):
+            dilation_layer = nn.Sequential(
+                DepthwiseSeparableConvModule(
+                    in_channels=(len(in_channels) - 1) * self.out_channels,
```
Comment (Junjun2016) [inline mmseg/models/necks/jpu.py:88]:
```suggestion
                    in_channels=(len(in_channels)) * self.out_channels,
```

File: mmseg/models/necks/jpu.py [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+            self.dilation_layers.append(dilation_layer)
+
+    def forward(self, inputs):
+        """Forward function."""
+        assert len(inputs) == len(self.in_channels), 'Length of inputs must \
+                                           be the same with self.in_channels!'
+
+        feats = [
+            self.conv_layers[i](inputs[i + 1])
+            for i in range(len(self.in_channels) - 1)
```
Comment (Junjun2016) [inline mmseg/models/necks/jpu.py:107]:
```suggestion
            self.conv_layers[i](inputs[i])
            for i in range(len(self.in_channels))
```

File: mmseg/models/necks/jpu.py [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+        ]
+
+        _, _, h, w = feats[0].size()
+        for i in range(len(feats) - 1):
```
Comment (Junjun2016) [inline mmseg/models/necks/jpu.py:111]:
```suggestion
        for i in range(1, len(feats)):
```

File: mmseg/models/necks/jpu.py [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+            feats[i + 1] = resize(
+                feats[i + 1],
```
Comment (Junjun2016) [inline mmseg/models/necks/jpu.py:113]:
```suggestion
            feats[i] = resize(
                feats[i],
```

File: mmseg/models/necks/jpu.py [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+                size=(h, w),
+                mode='bilinear',
+                align_corners=self.align_corners)
+
+        feat = torch.cat(feats[::-1], dim=1)
```
Comment (Junjun2016) [inline mmseg/models/necks/jpu.py:118]:
```suggestion
        feat = torch.cat(feats, dim=1)
```

File: mmseg/models/necks/jpu.py (hunk 6 above)
Comment (Junjun2016) [reply mmseg/models/necks/jpu.py:58]: `out_channels` --> `reduce_channels` or `middle_channels`.

File: mmseg/models/necks/jpu.py (hunk 12 above)
Comment (xvjiarui) [reply mmseg/models/necks/jpu.py:117]: Why reverse the order?

File: mmseg/models/necks/jpu.py [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+        feat = torch.cat([
+            self.dilation_layers[i](feat) for i in range(len(self.dilations))
+        ],
+                         dim=1)
+
+        outs = []
+        for i in range(len(inputs) - 1):
```
Comment (xvjiarui) [inline mmseg/models/necks/jpu.py:124]: Why `-1`?

File: mmseg/models/necks/jpu.py [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+            outs.append(inputs[i])
+        outs.append(feat)
+        outs = [outs[i] for i in self.out_indices]
```
Comment (xvjiarui) [inline mmseg/models/necks/jpu.py:127]: The usage of `self.out_indices` is not clear.

File: mmseg/models/necks/jpu.py (hunk 12 above)
Comment (MengzhangLI) [reply mmseg/models/necks/jpu.py:117]: Just to align metric of last version FastFCN, in my new version(not pushed yet) the order is not reversed.

File: mmseg/models/necks/jpu.py (hunk 13 above)
Comment (MengzhangLI) [reply mmseg/models/necks/jpu.py:124]: Because one largest feature map from backbone may be discarded for decoder head or auxiliary head. It could be kept although I think it is unnecessary.

File: mmseg/models/necks/jpu.py (hunk 14 above)
Comment (MengzhangLI) [reply mmseg/models/necks/jpu.py:127]: I follow similar usage in [fastscnn](https://github.com/open-mmlab/mmsegmentation/blob/36eb2d8bebcfca9b62ed8f27f9977dc0f12f87c8/mmseg/models/backbones/fast_scnn.py#L408).

File: mmseg/models/necks/jpu.py [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+                mode='bilinear',
+                align_corners=self.align_corners)
+
+        feat = torch.cat(feats, dim=1)
+        feat = torch.cat([
+            self.dilation_layers[i](feat) for i in range(len(self.dilations))
+        ],
+                         dim=1)
+
+        outs = []
```
Comment (xvjiarui) [inline mmseg/models/necks/jpu.py:123]: Add comment explain these.

Comment (xvjiarui) [reply mmseg/models/necks/jpu.py:123]:
```
return *inputs[:-1], feat
```

File: mmseg/models/necks/jpu.py [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+        h, w = feats[0].shape[2:]
+        for i in range(1, len(feats)):
+            feats[i] = resize(
+                feats[i],
+                size=(h, w),
+                mode='bilinear',
+                align_corners=self.align_corners)
+
+        feat = torch.cat(feats, dim=1)
+        feat = torch.cat([
```
Comment (xvjiarui) [inline mmseg/models/necks/jpu.py:118]: rename to `concat_feat`.

File: configs/_base_/models/fastfcn_r50-d32.py [hunk 17]
```diff
@@ -0,0 +1,52 @@
+# model settings
```
Comment (Junjun2016) [inline configs/_base_/models/fastfcn_r50-d32.py:1]: `configs/_base_/models/fastfcn_r50-d32.py`
--》
`configs/_base_/models/fastfcn_r50-d32_jpu_psp.py`

File: configs/fastfcn/fastfcn-jpu_deeplabv3_512x1024_80k_cityscapes.py [hunk 18]
```diff
@@ -0,0 +1,19 @@
+# model settings
```
Comment (Junjun2016) [inline configs/fastfcn/fastfcn-jpu_deeplabv3_512x1024_80k_cityscapes.py:1]: `configs/fastfcn/fastfcn-jpu_deeplabv3_512x1024_80k_cityscapes.py`
-->
`configs/fastfcn/fastfcn_r50-d32_jpu_aspp_512x1024_80k_cityscapes.py`

**Review by Junjun2016 — COMMENTED** [review]
Please rename all the config names accordingly.

File: configs/fastfcn/fastfcn-jpu_deeplabv3_512x1024_80k_cityscapes.py (hunk 18 above)
Comment (Junjun2016) [reply configs/fastfcn/fastfcn-jpu_deeplabv3_512x1024_80k_cityscapes.py:1]: include the README.md, model path, and log path.

File: configs/fastfcn/README.md [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+| ENCNet + JPU (4x4)| R-50-D32  | 512x1024  |   80000 | 15.45        | -              | 78.6 |         80.25 | [config](https://github.com/open-mmlab/mmsegmentation/blob/master/configs/fastfcn/fastfcn_r50-d32_jpu_enc_4x4_512x1024_80k_cityscapes.py)  | [model](https://download.openmmlab.com/mmsegmentation/v0.5/fastfcn/fastfcn_r50-d32_jpu_enc_4x4_512x1024_80k_cityscapes/fastfcn_r50-d32_jpu_enc_4x4_512x1024_80k_cityscapes_20210926_093217-e1eb6dbb.pth) &#124; [log](https://download.openmmlab.com/mmsegmentation/v0.5/fastfcn/fastfcn_r50-d32_jpu_enc_4x4_512x1024_80k_cityscapes/fastfcn_r50-d32_jpu_enc_4x4_512x1024_80k_cityscapes_20210926_093217.log.json)     |
```
Comment (Junjun2016) [inline configs/fastfcn/README.md:36]: ENCNet --》 EncNet

File: configs/fastfcn/README.md [hunk 20]
```diff
@@ -0,0 +1,41 @@
...
+| EncNet + JPU (4x4)| R-50-D32  | 512x1024  |   80000 | 15.45        | -              | 78.6 |         80.25 | [config](https://github.com/open-mmlab/mmsegmentation/blob/master/configs/fastfcn/fastfcn_r50-d32_jpu_enc_4x4_512x1024_80k_cityscapes.py)  | [model](https://download.openmmlab.com/mmsegmentation/v0.5/fastfcn/fastfcn_r50-d32_jpu_enc_4x4_512x1024_80k_cityscapes/fastfcn_r50-d32_jpu_enc_4x4_512x1024_80k_cityscapes_20210926_093217-e1eb6dbb.pth) &#124; [log](https://download.openmmlab.com/mmsegmentation/v0.5/fastfcn/fastfcn_r50-d32_jpu_enc_4x4_512x1024_80k_cityscapes/fastfcn_r50-d32_jpu_enc_4x4_512x1024_80k_cityscapes_20210926_093217.log.json)     |
+
+Note:
+
+- `4x4` means 4 GPUs with 4 samples per GPU in training, default setting is 4 GPUs with 2 samples per GPU in training.
+- Results of [DeepLabV3 (mIoU: 79.32)](https://github.com/open-mmlab/mmsegmentation/blob/master/configs/deeplabv3), [PSPNet (mIoU: 78.55)](https://github.com/open-mmlab/mmsegmentation/blob/master/configs/pspnet) and [ENCNet (mIoU: 77.94)](https://github.com/open-mmlab/mmsegmentation/blob/master/configs/encnet) can be found in each original repo.
```
Comment (Junjun2016) [inline configs/fastfcn/README.md:41]:
```suggestion
- Results of [DeepLabV3 (mIoU: 79.32)](https://github.com/open-mmlab/mmsegmentation/blob/master/configs/deeplabv3), [PSPNet (mIoU: 78.55)](https://github.com/open-mmlab/mmsegmentation/blob/master/configs/pspnet) and [ENCNet (mIoU: 77.94)](https://github.com/open-mmlab/mmsegmentation/blob/master/configs/encnet) can be found in each original repository.
```

**Review by Junjun2016 — APPROVED** [review]
LGTM, please also resolve the comment.

File: mmseg/models/necks/jpu.py [hunk 21]
```diff
@@ -0,0 +1,127 @@
...
+        else:
+            self.backbone_end_level = end_level
+            assert end_level <= len(in_channels)
+
+        self.dilations = dilations
+        self.align_corners = align_corners
+
+        self.conv_layers = nn.ModuleList()
+        self.dilation_layers = nn.ModuleList()
+        for i in range(self.start_level, len(in_channels)):
```
Comment (xvjiarui) [inline mmseg/models/necks/jpu.py:72]:
```suggestion
        for i in range(self.start_level, self.backbone_end_level):
```

File: mmseg/models/necks/jpu.py [hunk 22]
```diff
@@ -0,0 +1,127 @@
...
+                    pw_norm_cfg=norm_cfg,
+                    pw_act_cfg=act_cfg))
+            self.dilation_layers.append(dilation_layer)
+
+    def forward(self, inputs):
+        """Forward function."""
+        assert len(inputs) == len(self.in_channels), 'Length of inputs must \
+                                           be the same with self.in_channels!'
+
+        feats = [self.conv_layers[i](inputs[i]) for i in range(len(inputs))]
```
Comment (xvjiarui) [inline mmseg/models/necks/jpu.py:103]: Should use start end level. 
Please refer to fpn.py

File: mmseg/models/necks/jpu.py [hunk 23]
```diff
@@ -0,0 +1,127 @@
...
+
+        outs = []
+
+        # Default: outs[2] is the output of JPU for decoder head, outs[1] is
+        # the feature map from backbone for auxiliary head. Additionally,
+        # outs[0] can also be used for auxiliary head.
+        for i in range(len(inputs) - 1):
+            outs.append(inputs[i])
+        outs.append(concat_feat)
+        return tuple(outs)
```
Comment (xvjiarui) [inline mmseg/models/necks/jpu.py:127]: Not all input will be outputed. If start end level is not default.

Comment (xvjiarui) [reply mmseg/models/necks/jpu.py:127]: Please also add unittest for these cases.

**Review by xvjiarui — APPROVED** [review]
LGTM. Please resolve the conflict.
